### DISCUSSION TITLE
[1 febbraio 2022 - A] Esercizio 6

### DISCUSSION THREAD

**Discussion by matypist** [sapienzastudentsnetwork/psd2223#210] [Esami]
**Esercizio 6 (5 punti):**

Si consideri il numero esadecimale $4A5B$ e gli si sottragga in base 16 il numero esadecimale $6C1$. Si converta poi il risultato in una sequenza binaria di 16 bit, da interpretarsi come un numero razionale in formato IEEE half-precision. Al numero così ottenuto si sottragga il numero con $1100011110000000_2$ espresso sempre in formato IEEE half-precision, e si scriva il risultato nello stesso formato.

**Comment by Elia-Belli** [comment]
- $4A5B -$
  $06C1 =$
  $439A_{16} = 0\ 10000\ 1110011010 = +2^1\ 1.111001101 = +2^2\ 0.1111001101= X$

- $Y = 1\ 10001\ 1110000000 = -2^2\ 1.111$
  $(X-(-Y)) = (X+Y)$ quindi non c'è bisogno di passare al complemento

- $00.1111001101 +$
  $01.1110000000 =$
  $10.1101001101$
$Z = 2^2\ 10.1101001101 = 2^3\ 1.01101001101 = 0\ 10010\ 0110100110$ (ho normalizzato troncando la cifra meno significativa)

**Comment by Baliano1999** [comment]
Scusate, ma non mi torna una cosa: 

Il numero che ci siamo appena calcolati 439A è un numero che, convertito in virgola mobile deve essere rappresentato nello standard IEEE 754 SOMMANDO il bias. 

Perchè in questo caso è stato sottratto?

**Comment by Elia-Belli** [comment]
> Scusate, ma non mi torna una cosa:
> 
> Il numero che ci siamo appena calcolati 439A è un numero che, convertito in virgola mobile deve essere rappresentato nello standard IEEE 754 SOMMANDO il bias.
> 
> Perchè in questo caso è stato sottratto?

- Stiamo decodificando un numero da IEEE754 a natazione scientifica binaria, quindi all'esponente va sottratto il bias $16-15 = 1$.
- Il Bias ( $2^4-1=15$ ) è appositamente utilizzato per rappresentare sia esponenti sia negativi che positivi in codifica **unsigned binary**, esempio:
sia l'esponente da rappresentare $-12$, allora applichiamo il bias $-12+15=3$ (positivo), adesso nei 5 bit dell'esponente inseriamo la codifica senza segno di $3 \rightarrow 00011$; se andiamo a decodificare l'esponente $00011\rightarrow 3$, però avevamo codificato un numero negativo, infatti dobbiamo applicare il bias $3-15=-12$ ed ecco fatto. (se nella decodifica sommassimo anzichè sottrarre non otterremo mai valori negativi)
